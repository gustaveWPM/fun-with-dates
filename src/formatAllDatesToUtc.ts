type _DistributeDatesToStrings<T> = {
  [K in keyof T]: DatesToStrings<T[K]>;
};

type DatesToStrings<T> = T extends Date ? string : T extends object ? _DistributeDatesToStrings<T> : T;

const formatUtcDate = (d: Date): string => `implement this ${d}`;

function formatEntry<Input extends object>(k: keyof Input, v: Input[typeof k]) {
  if (v instanceof Date) return [k, formatUtcDate(v)];
  if (v && typeof v === 'object') return [k, formatAllDatesToUtc(v)];
  return [k, v];
}

export const formatAllDatesToUtc = <Input extends object>(input: Input): DatesToStrings<Input> =>
  Object.fromEntries(Object.entries(input).map(([k, v]) => formatEntry(k, v)));
