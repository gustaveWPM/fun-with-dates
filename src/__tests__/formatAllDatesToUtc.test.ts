import { describe, expect, it } from 'vitest';

import { formatAllDatesToUtc } from '../formatAllDatesToUtc';
import DATASET from './data/dataset';

describe('formatAllDatesToUtc [impl]', () => {
  it('should let the structure unchanged, given an empty structure', () => {
    const input = {};
    const output = formatAllDatesToUtc(input);

    expect(output).toStrictEqual(input);
  });

  it('should let the structure unchanged, given a flat structure without dates', () => {
    const input = DATASET.flatStructureWithoutDates;
    const output = formatAllDatesToUtc(input);

    expect(output).toStrictEqual(input);
  });

  it('should preserve functions, given a structure with a function', () => {
    const input = DATASET.flatStructureWithoutDates;
    const output = formatAllDatesToUtc(input);

    // eslint-disable-next-line no-magic-numbers
    const [string, number] = ['string', 0];
    const obj = { string, number };

    expect(output.fun(string, number, obj)).toBe(input.fun(string, number, obj));
  });

  it('should replace all dates by formatted strings, given a flat structure with dates', () => {
    const input = DATASET.flatStructureWithDates;
    const output = formatAllDatesToUtc(input);

    expect(output).toMatchSnapshot();
  });

  it('should let the structure unchanged, given a structure with nestings & without dates', () => {
    const input = DATASET.deepStructureWithoutDates;
    const output = formatAllDatesToUtc(input);

    expect(output).toStrictEqual(input);
  });

  it('should recursively replace all dates by formatted strings, given a structure with dates & nestings', () => {
    const input = DATASET.deepStructureWithDates;
    const output = formatAllDatesToUtc(input);

    expect(output).toMatchSnapshot();
  });
});
