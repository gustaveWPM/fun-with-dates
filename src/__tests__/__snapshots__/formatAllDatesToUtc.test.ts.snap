// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`formatAllDatesToUtc [impl] > should recursively replace all dates by formatted strings, given structure with dates & nestings 1`] = `
{
  "bar": "bar",
  "date": "1998-02-01T01:00:00+01:00",
  "date2": "1998-01-02T01:00:00+01:00",
  "foo": "foo",
  "nested": {
    "date": "1998-02-01T01:00:00+01:00",
    "deeplyNested": {
      "12": 12,
      "date": "1998-02-01T01:00:00+01:00",
      "foo": "foo",
    },
    "foo": "foo",
  },
}
`;

exports[`formatAllDatesToUtc [impl] > should replace all dates by formatted strings, given flat structure with dates 1`] = `
{
  "bar": "bar",
  "date": "1998-02-01T01:00:00+01:00",
  "date2": "1998-01-02T01:00:00+01:00",
  "foo": "foo",
}
`;
